From d30b8b83175fbf96644cffbda37e90cb4703c139 Mon Sep 17 00:00:00 2001
From: Todd Lipcon <todd@cloudera.com>
Date: Thu, 12 May 2011 15:07:51 -0700
Subject: [PATCH 0960/1179] HADOOP-6947. Kerberos login should set the refreshKrb5Config option

Reason: necessary for daemons that will use multiple keytab files for different principals
Author: Todd Lipcon
Ref: CDH-3184
---
 .../hadoop/security/UserGroupInformation.java      |    1 +
 .../hadoop/security/ManualTestKeytabLogins.java    |   57 ++++++++++++++++++++
 2 files changed, 58 insertions(+), 0 deletions(-)
 create mode 100644 src/test/org/apache/hadoop/security/ManualTestKeytabLogins.java

diff --git a/src/core/org/apache/hadoop/security/UserGroupInformation.java b/src/core/org/apache/hadoop/security/UserGroupInformation.java
index df2bf31..018c724 100644
--- a/src/core/org/apache/hadoop/security/UserGroupInformation.java
+++ b/src/core/org/apache/hadoop/security/UserGroupInformation.java
@@ -392,6 +392,7 @@ public class UserGroupInformation {
       KEYTAB_KERBEROS_OPTIONS.put("doNotPrompt", "true");
       KEYTAB_KERBEROS_OPTIONS.put("useKeyTab", "true");
       KEYTAB_KERBEROS_OPTIONS.put("storeKey", "true");
+      KEYTAB_KERBEROS_OPTIONS.put("refreshKrb5Config", "true");
       KEYTAB_KERBEROS_OPTIONS.putAll(BASIC_JAAS_OPTIONS);      
     }
     private static final AppConfigurationEntry KEYTAB_KERBEROS_LOGIN =
diff --git a/src/test/org/apache/hadoop/security/ManualTestKeytabLogins.java b/src/test/org/apache/hadoop/security/ManualTestKeytabLogins.java
new file mode 100644
index 0000000..632ceaf
--- /dev/null
+++ b/src/test/org/apache/hadoop/security/ManualTestKeytabLogins.java
@@ -0,0 +1,57 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.security;
+
+import org.apache.hadoop.security.UserGroupInformation;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Regression test for HADOOP-6947 which can be run manually in
+ * a kerberos environment.
+ *
+ * To run this test, set up two keytabs, each with a different principal.
+ * Then run something like:
+ *  <code>
+ *  HADOOP_CLASSPATH=build/test/classes bin/hadoop \
+ *     org.apache.hadoop.security.ManualTestKeytabLogins \
+ *     usera/test@REALM  /path/to/usera-keytab \
+ *     userb/test@REALM  /path/to/userb-keytab
+ *  </code>
+ */
+public class ManualTestKeytabLogins {
+
+  public static void main(String []args) throws Exception {
+    if (args.length != 4) {
+      System.err.println(
+        "usage: ManualTestKeytabLogins <principal 1> <keytab 1> <principal 2> <keytab 2>");
+      System.exit(1);
+    }
+
+    UserGroupInformation ugi1 =
+      UserGroupInformation.loginUserFromKeytabAndReturnUGI(
+        args[0], args[1]);
+    System.out.println("UGI 1 = " + ugi1);
+    assertTrue(ugi1.getUserName().equals(args[0]));
+    
+    UserGroupInformation ugi2 =
+      UserGroupInformation.loginUserFromKeytabAndReturnUGI(
+        args[2], args[3]);
+    System.out.println("UGI 2 = " + ugi2);
+    assertTrue(ugi2.getUserName().equals(args[2]));
+  }
+}
-- 
1.7.0.4

